Feature/longbench v2 evaluation utils

alhridoy opened the pull request:
Fixes #10921


 Summary
 This PR implements evaluation utilities for the LongBench-v2 dataset,
 enabling SGLang users to easily evaluate models on long-context
 understanding tasks.


 Implementation Details
  - `simple_eval_longbench_v2.py`: Main evaluation utility following
    SGLang patterns (similar to simple_eval_gpqa.py,
    simple_eval_mmlu.py)
  - `test_longbench_v2_eval.py`: Comprehensive test suite

alhridoy committed:
feat: Add LongBench-v2 evaluation utilities for SGLang

- Implement comprehensive LongBench-v2Eval class with support for all 6 task categories
- Add dataset loading from HuggingFace datasets and local files (JSON/CSV/JSONL)
- Support category filtering, context length filtering, and parallel processing
- Integrate with SGLang evaluation framework via run_eval.py
- Add comprehensive test suite with unit tests and integration tests
- Include detailed documentation with usage examples and configuration options
- Support metrics tracking by category and context length ranges

Features:
- Single-document QA, Multi-document QA, Long in-context learning
- Long-dialogue history, Code repo understanding, Long structured data
- Context lengths from 8k to 2M words with intelligent filtering
- Compatible with existing SGLang SamplerBase and evaluation patterns

alhridoy committed:
feat: Add LongBench-v2 evaluation utility

- Implement simple_eval_longbench_v2.py following SGLang patterns
- Add official LongBench-v2 prompt template and answer extraction
- Support for domain-based categorization and metrics
- Compatible with official LongBench-v2 dataset format
- Includes comprehensive test suite

Addresses feature request #10921

Fridge003 commented:
@alhridoy Please fix lint with 
```bash
pip3 install pre-commit
pre-commit install
pre-commit run --all-files
```

Fridge003 commented on docs/longbench_v2_evaluation.md:
```diff
@@ -0,0 +1,217 @@
+# LongBench-v2 Evaluation Guide
```
Can you move this document to `python/sglang/test`? The documents under docs are supposed to be shown on official document site, while this document is more for developer testing.

alhridoy committed:
Resolve merge conflict: integrate LongBench-v2 and MMMU evaluations

- Fixed LongBench-v2 dataset field mapping (A/B/C/D, category)
- Added HuggingFace dataset loading and filtering capabilities
- Moved documentation to python/sglang/test/ directory
- Integrated MMMU evaluation from main branch
- Updated constructor interface to use data_source parameter
- Applied formatting fixes from pre-commit hooks

Fridge003 commented:
@alhridoy Can you post some accuracy results on evaluation with the newly added scripts, and compare them with official LongBench results?

alhridoy commented:
> @alhridoy Can you post some accuracy results on evaluation with the newly added scripts, and compare them with official LongBench results?

Summary (20-sample spot check):
     • Overall accuracy: 55.0%
     • Multi-Document QA: 3/5 (60%)
     • Long Structured Data Understanding: 2/3 (66.7%)
     • Single-Document QA: 2/6 (33.3%)
     • Code Repository Understanding: 2/2 [(100%)](url)
     • Long-dialogue History: 1/3 (33.3%)
     • Long In-context Learning: 1/1 (100%)
     • Context buckets — short: 1/1 (100%), medium: 6/11 (54.5%), long: 4/8 (50%)

alhridoy committed:
feat: Enhanced LongBench-v2 evaluation with HuggingFace/local support

- Updated simple_eval_longbench_v2.py to load/normalize official LongBench-v2 schema
- Added support for HuggingFace datasets and local CSV/JSON/JSONL files
- Improved answer extraction and choice handling with fallbacks
- Refreshed test_longbench_v2_eval.py to use JSON fixtures and removed pandas dependency
- Added executable validation scripts (validate_longbench_v2.py and _standalone.py)
- Enhanced metrics and category filtering capabilities

alhridoy committed:
fix: Update CLI default dataset from zai-org to official THUDM/LongBench-v2

- Updated run_eval.py default data_source
- Updated simple_eval_longbench_v2.py default data_source parameter
- Updated all documentation examples to use official dataset

alhridoy committed:
resolve: Merge upstream/main with LongBench-v2 evaluation

- Resolved merge conflict in run_eval.py
- Preserved LongBench-v2 evaluation functionality with official THUDM dataset
- Integrated upstream MMMU evaluation and return_latency features
- Maintained all CLI arguments for LongBench-v2 evaluation

Fridge003 commented on python/sglang/test/longbench_v2_evaluation.md:
```diff
@@ -0,0 +1,217 @@
+- `short_context`: Accuracy on contexts < 32k tokens
+- `medium_context`: Accuracy on contexts 32k-128k tokens
+- `long_context`: Accuracy on contexts > 128k tokens
+
```
In original LongBench-v2 dataset, the examples are divided into "easy" ones and "hard" ones. Is it possible for we to generate metrics according to difficutly?

Fridge003 commented on python/sglang/test/run_eval.py:
```diff
@@ -60,6 +60,23 @@ def run_eval(args):
+        from sglang.test.simple_eval_longbench_v2 import LongBenchV2Eval
+
+        # Default to HuggingFace dataset, can be overridden with --dataset-path
+        data_source = getattr(args, "dataset_path", "THUDM/LongBench-v2")
```
How about setting default values below, in `parser.add_argument` part

Fridge003 commented on python/sglang/test/simple_eval_longbench_v2.py:
```diff
@@ -0,0 +1,331 @@
+        rng = random.Random(0)
+        if num_examples:
+            assert n_repeats == 1, "n_repeats only supported when not sampling examples"
+            examples = rng.sample(examples, min(num_examples, len(examples)))
```
Can we just pick the first `num_examples`  examples rather than randomly picking? Since when we do benchmark we wish the input are the same set of prompts.

Fridge003 commented on python/sglang/test/simple_eval_longbench_v2.py:
```diff
@@ -0,0 +1,331 @@
+        for example in examples:
+            context = example.get("context", "")
+            # Rough token count approximation (4 chars per token)
+            token_count = len(context) // 4
```
How about setting max_length and min_length as the range of input string length.
Since the estimation of token count is not accurate

Fridge003 commented:
@alhridoy I feel we can create a new folder `longbench_v2` under `python/sglang/test`, and move all the newly created files except `simple_eval_longbenchv2.py` to that folder.

alhridoy committed:
refine: address LongBench-v2 evaluation feedback

Fridge003 committed:
Merge branch 'main' into feature/longbench-v2-evaluation-utils

Fridge003 committed:
fix lint

Fridge003 reviewed (approved):
LGTM

ispobock reviewed (approved):
LGTM
